File: ruby/beer-song/beer_song.rb
class BeerSong

  attr_reader :current_bottle

  def verses(start, stop = nil)
    stop = start if stop.nil?
    (stop..start).to_a.reverse.map do |x|
      verse(x)
    end.join("\n")
  end

  def verse(n)
    self.class.verse_factory(n).verse
  end

  def lyrics
    verses(99, 0)
  end

  def self.verse_factory(n)
    case n
    when 0
      return VerseZero.new(n)
    when 1
      return VerseOne.new(n)
    when 99
      return Verse99.new(n)
    else
      return VerseX.new(n)
    end
  end

  class VerseX
    def initialize(n)
      @num = n
      @oneit = 'one'
      @bottles = 'bottles'
      @next_verse = n - 1
    end

    attr_reader :num, :oneit, :bottles, :next_verse

    def verse
      first_line + BeerSong
        .verse_factory(next_verse).second_line
    end

    def first_line
      "#{the_wall}, #{num_beers.downcase}.\n"
    end

    def  second_line
      "#{action}, #{the_wall.downcase}.\n"
    end

    def action
      "Take #{oneit} down and pass it around"
    end

    def num_beers
      "#{num} #{bottles} of beer"
    end

    def the_wall
      "#{num_beers} on the wall"
    end
  end

  class Verse99 < VerseX
    def action
      "Go to the store and buy some more"
    end
  end

  class VerseOne < VerseX
    def initialize(n = 1)
      super(n)
      @bottles = 'bottle'
    end
  end

  class VerseZero < VerseX
    def initialize(n = 0)
      super(n)
      @next_verse = 99
      @oneit = 'it'
    end

    def num_beers
      "No more #{bottles} of beer"
    end

  end


end

module BookKeeping
  VERSION = 2
end

File: javascript/beer-song/beer-song.js
var BeerSong = function () {

  var bottleCount = function(num) {
    if (num == 0) {return 'No more';}
    if (num < 0){ return "99".toString();}
    return num.toString();
  };

  var action = function(num) {
    if (num == 0) return "Go to the store and buy some more, "
    return "Take " + (num == 1 ? 'it' : 'one') + " down and pass it around, ";
  };

  var bottle = function(num) {
    return num == 1 ? 'bottle' : 'bottles';
  }

  this.verse = function(num) {
    return bottleCount(num) +  " " + bottle(num) + " of beer on the wall, " +
    bottleCount(num).toLowerCase() +  " " + bottle(num) + " of beer.\n" +
    action(num) +
    bottleCount(num - 1).toLowerCase() + " " + bottle(num-1) + " of beer on the wall.\n";
  }

  this.sing = function(start, stop) {
    var song = [];
    if (stop === undefined) stop = 0;
    for (var i = start; i >= stop; i--) {
      song.push(this.verse(i));
    }
    return song.join("\n");
  }
};

module.exports = BeerSong;
